src/mongo_odm/query_builder.hpp

```diff
@@ -25,93 +26,453 @@
 #include <bsoncxx/view_or_value.hpp>
 
 #include <mongo_odm/nvp.hpp>
+#include <mongo_odm/util.hpp>
 
 namespace mongo_odm {
 MONGO_ODM_INLINE_NAMESPACE_BEGIN
 
-// Forward declarations
-template <typename Base, typename T>
-class NotExpr;
+// Forward declarations for Expression type trait structs
+template <typename>
+struct is_sort_expression;
+template <typename>
+struct is_query_expression;
+template <typename>
+struct is_update_expression;
+
+/**
+ * Type trait that contains true if a type T can be appended to a BSON builder using
+ * builder.append(T val);
+ */
+template <typename T>
+constexpr auto is_bson_appendable_impl(int)
```

Comment:
@Machyne it seems like without the `int` argument, the two function definitions are ambiguous and it fails to compile.